### ISSUE TITLE
Closing terminal while WSL nodejs is running keeps processes alive and may cause excessive CPU load

### ISSUE THREAD

**Issue by CherryDT** [Microsoft/vscode#44408]
For the last month or so, I am experiencing an issue with the integrated terminal which is very annoying. I'm using WSL bash as shell, and whenever node is running inside the shell, I will get issues when the terminal is closed without pressing Ctrl+C first (which can happen in various ways, some more subtle than others, for example switching projects or reloading the window after installing an extension - but of course it also happens when clicking the trash bin icon).

I'm not sure yet who is at fault here, maybe it's not VSCode's fault at all, but I can't be certain.

* It does not happen in a regular console window outside VSCode. When I close such a window, the whole process tree is terminated as I would expect.
* Until now I only saw this behavior with nodejs. I also tried vim but that one was nicely closed.
* It used to work in the past.

This has two consequences:

* The whole process tree including the shell remain alive, but with their terminal shown as `?` in `ps`. This already caused weird issues for me multiple times because a nodejs process which I had forgotten was still running invisibly in the background and doing things like locking files, serving an outdated version of a website, blocking a port, things like that. Remember that this also affects any kind of package which is actually a node application, for example `http-server`.
    * When the nodejs process was killed, the shell process will also go away.
* Worse than that, when the nodejs process is listening for console input (e.g. just start `node` to get to the REPL prompt), it seems like it starts eating CPU after it got into that "detached" state. In that state, `kill` also won't kill it any longer, I need `kill -9`. I'm not yet sure why it does that, though - maybe it's a bug in node when you remove the terminal "carpet" below its feet?

![image](https://user-images.githubusercontent.com/6010463/36667299-b181b804-1aed-11e8-9c83-c1c14f2612c4.png)

Version 1.21.0-insider
Commit 903169966590cd236c5016cdc774875215d31765
Date 2018-02-26T05:18:08.801Z
Shell 1.7.9
Renderer 58.0.3029.110
Node 7.9.0 (in WSL: 8.4.0)
Architecture x64
Also happens with extensions disabled.
Also happens with 32-bit version.
Also happens with 1.15.0-insider.

PS: I know there is https://github.com/Microsoft/WSL/issues/2496 but the symptoms don't quite match...

**Comment by Tyriar** [comment]
@CherryDT can you get [ProcessExplorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) and take a screenshot of the problem process trees before and after the issue?

Here's how we clean up afterwards: https://github.com/Tyriar/node-pty/blob/ad71e8989d7613e2e439cc191ede6d5e48600f58/src/windowsPtyAgent.ts#L94-L105 Maybe WSL processes are treated differently somehow, would be interested to see the process tree 😃

**Comment by CherryDT** [comment]
The tree looks the same, and it isn't a tree in the first place.

This is because the parent of `bash` is a non-existent process, also `node`'s. Don't ask me why.

However, `pstree` in Linux _does_ show a parent-child relationship:

```init─┬─init───bash───node─┬─4*[{V8 WorkerThread}]
     │                    └─5*[{node}]
     └─init───bash───pstree```

When looking at the "Log" in Process Explorer, it shows it like this:

```00:28:46 02.03.2018: Process created: Code - Insiders.exe (13924) started by Code - Insiders.exe (61896)
00:28:46 02.03.2018: Process created: winpty-agent.exe (66304) started by Code - Insiders.exe (13924)
00:28:46 02.03.2018: Process created: conhost.exe (64408) started by winpty-agent.exe (66304)
00:28:46 02.03.2018: Process created: bash.exe (51012) started by winpty-agent.exe (66304)
00:28:46 02.03.2018: Process created: init (27184) started by init (16892)
00:28:46 02.03.2018: Process created: bash (52840) started by Unknown process (0)
00:28:46 02.03.2018: Process created: bash (21820) started by bash (52840)
00:28:47 02.03.2018: Process terminated: bash (21820); exit status 0x0
00:28:48 02.03.2018: Process created: node (37056) started by Unknown process (0)
```

**Comment by Tyriar** [comment]
@bitcrazed does the WSL process not live underneath the process that launched it? 😕

**Comment by CherryDT** [comment]
Why was it closed? What more information is needed?

**Comment by bitcrazed** [comment]
Hey @CherryDT - thanks for your detailed analysis here. 

@Tyriar - does VSCode kill the WSL instance the integrated terminal is closed by the user?

I was running cacafire in Ubuntu under WSL:
![image](https://user-images.githubusercontent.com/961950/37549939-c159d55a-2942-11e8-8f84-b4bdb2f467c7.png)

When I closed the terminal, it appears that cacafire continues to run in the WSL instance that should've been killed when the terminal was closed:
![image](https://user-images.githubusercontent.com/961950/37549946-e1d65ba0-2942-11e8-9752-436a54739098.png)

**Comment by Tyriar** [comment]
> does VSCode kill the WSL instance the integrated terminal is closed by the user?

@bitcrazed yes, we kill the pty and then the *console* process list here:

https://github.com/Tyriar/node-pty/blob/ad71e8989d7613e2e439cc191ede6d5e48600f58/src/windowsPtyAgent.ts#L92-L105

Any advice on the right way to do this would be helpful.

**Comment by alexr00** [comment]
I ran node from within WSL in the integrated terminal then killed the terminal. I didn't see any orphaned processes. Looks like this is fixed by 95c0e0728b4

**Comment by CherryDT** [comment]
Confirmed!
